Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/IndirectILLEnergyTransfer.py

```diff
@@ -502,26 +512,44 @@ def _convert_to_energy_bats(self, ws, epp_ws, t0_offset=0.):
         output_epp = self.getPropertyValue('OutputElasticChannelWorkspace')
         rows = epp_ws.rowCount()
-        if rows == 1:
-            elastic_channel_equator = epp_ws.cell('PeakCentre', 0)
-            formula = '{0} + (x - {1})*{2}'.format(elastic_tof_equator, elastic_channel_equator, channel_width)
-            ConvertAxisByFormula(InputWorkspace=ws, Axis='X', AxisUnits='TOF', Formula=formula, OutputWorkspace=ws)
-        else:
-            elastic_channel_equator = epp_ws.cell('PeakCentre', rows-1)
-            x = ws.extractX()
-            for pixel in range(ws.getNumberHistograms()):
-                if 0 < pixel < N_PIXELS_PER_TUBE * N_TUBES + 1:
-                    group = int((pixel - 1) / self._group_by)
-                    if epp_ws.cell('FitStatus', group) == 'success':
-                        l2 = detector_info.l2(pixel)
-                        elastic_tof = ((l1 + l2) / v_fixed + t0_offset) * 1E+6
-                        elastic_channel = epp_ws.cell('PeakCentre', group)
-                else:
-                    elastic_tof = elastic_tof_equator
-                    elastic_channel = elastic_channel_equator
-                x_new = elastic_tof + (x[pixel] - elastic_channel) * channel_width
-                ws.setX(pixel, x_new)
-            ws.getAxis(0).setUnit('TOF')
+
+        elastic_channel_equator = epp_ws.cell('PeakCentre', rows-1)
+        x = ws.extractX()
+
+        elastic_tof = elastic_tof_equator
+        elastic_channel = elastic_channel_equator
+
+        x_new = elastic_tof + (x[0] - elastic_channel) * channel_width
+        ws.setX(0, x_new)
+
+        for pixel in range(1, N_PIXELS_PER_TUBE * N_TUBES + N_MONITOR):
+
+            group = (pixel - 1) // self._group_by
+            if self._fit_option == "FitAllPixelGroups" and epp_ws.cell('FitStatus', group) == 'success':
+                l2 = detector_info.l2(pixel)
+                elastic_tof = ((l1 + l2) / v_fixed + t0_offset) * 1E+6
+                elastic_channel = epp_ws.cell('PeakCentre', group)
+            else:
+                elastic_tof = elastic_tof_equator
+                elastic_channel = elastic_channel_equator
+            x_new = elastic_tof + (x[pixel] - elastic_channel) * channel_width
+            ws.setX(pixel, x_new)
+
+        for single_detector in range(N_PIXELS_PER_TUBE * N_TUBES + N_MONITOR, ws.getNumberHistograms()):
+            group = rows - 1 - (ws.getNumberHistograms() - single_detector)
+            if epp_ws.cell('FitStatus', group) in ['success', "narrowPeak"]:
+                l2 = detector_info.l2(single_detector)
+                elastic_tof = ((l1 + l2) / v_fixed + sd_t0_offset) * 1E+6
+                elastic_channel = epp_ws.cell('PeakCentre', group)
+            else:
+                elastic_tof = elastic_tof_equator
+                elastic_channel = elastic_channel_equator
+                self.log.warning("Single detector number {0}'s peak was not found. Using equatorial peak instead."
```

Comment:
In this case could we not calculate a guess based on the relative distance and relative delay differences of the PSD equator and SD?

Reply:
This feature is not urgent, so I created a separate issue to address it later: https://github.com/mantidproject/mantid/issues/31399